### ISSUE TITLE
Model Card / Model output discrepancy for tflite iris model

### ISSUE THREAD

**Issue by shortcipher3** [google/mediapipe#1041]
I am trying to understand the output of the tflite Iris landmarks available here: https://github.com/google/mediapipe/tree/master/mediapipe/models

The model card:
https://mediapipe.page.link/iris-mc
describes the output as 71 2D landmarks and 5 2D landmarks.  When I perform inference with tensorflow lite, I get 213 values and 15 values in the model outputs - so I assume I am getting an x/y/z coordinate for each point - however the values are in the -7000 to +7000 range.  My input was a 64x64 image, any idea of how these points correspond to the original image?

The tensorflow model's output defaults to float32, should it be some other format?

**Comment by fanzhanggoogle** [comment]
Yes, 213 values are x,y,z for 71 refined eye contour landmarks and 15 values are x,y,z for 5 iris landmarks. x,y is normalized by input image size, so should be within [0, 1]. z for refined eye contour landmarks is the same scale as face landmarks. You can check their correspondence to face landmarks [here](https://github.com/google/mediapipe/blob/8f69af91feb5944e130959dc9cc256b93d03abd2/mediapipe/graphs/iris_tracking/calculators/update_face_landmarks_calculator.cc#L33). z for iris landmarks is not trained with ground truth, but with average of eye contour landmarks. So should be in the same scale, -7000 to +7000 doesn't make sense to me. Have you tried Mediapipe inference pipeline for iris model for a sanity check?

**Comment by shortcipher3** [comment]
@fanzhanggoogle  thanks for the explanation.

After digging deeper into the code I found that my problem with the range was two-fold, the input image pixel values to the tflite model are first normalized to the range 0-1 (by the `tflite_converter_calculator` node), and the output landmarks are normalized to the range 0-1 by dividing by the image width and height (by the `tflite_tensors_to_landmarks_calculator` node).

The model card, I assume is for the tflite model?   To correct the model card in the input it should specify the normalization, for the output it should specify the outputs are 3D not 2D and specify the normalization required.

If others are interested in working python code for running the tflite model by itself here it is (tested in colab):

```
!wget https://github.com/google/mediapipe/raw/master/mediapipe/models/iris_landmark.tflite
import tensorflow as tf
import numpy as np

model_file = 'iris_landmark.tflite'
interpreter = tf.lite.Interpreter(
    model_path=model_file)
interpreter.allocate_tensors()

input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

print(f'input_details {input_details}')
print(f'output_details {output_details}')

# check the type of the input tensor
floating_model = input_details[0]['dtype'] == np.float32
print(f'is floating model: {floating_model}')

# NxHxWxC, H:1, W:2
height = input_details[0]['shape'][1]
width = input_details[0]['shape'][2]
img = np.random.randint(0, 256, (1,64,64,3), np.uint8)

if floating_model:
  img = (np.float32(img) - 0.0) / 255.0 # normalization from tflite_converter_calculator

interpreter.set_tensor(input_details[0]['index'], img)

interpreter.invoke()

output_eyes = interpreter.get_tensor(output_details[0]['index'])[0]
output_iris = interpreter.get_tensor(output_details[1]['index'])[0]

print(f'output iris landmarks x {output_iris[::3]/64.0}') # tflite_tensors_to_landmarks_calculator
print(f'output iris landmarks y {output_iris[1::3]/64.0}')
print(f'output iris landmarks z {output_iris[2::3]}')

print(f'output eye landmarks x {output_eyes[::3]/64.0}')
print(f'output eye landmarks y {output_eyes[1::3]/64.0}')
print(f'output eye landmarks z {output_eyes[2::3]}')
```

**Comment by sandeep22-v** [comment]
> @fanzhanggoogle thanks for the explanation.
> 
> After digging deeper into the code I found that my problem with the range was two-fold, the input image pixel values to the tflite model are first normalized to the range 0-1 (by the `tflite_converter_calculator` node), and the output landmarks are normalized to the range 0-1 by dividing by the image width and height (by the `tflite_tensors_to_landmarks_calculator` node).
> 
> The model card, I assume is for the tflite model? To correct the model card in the input it should specify the normalization, for the output it should specify the outputs are 3D not 2D and specify the normalization required.
> 
> If others are interested in working python code for running the tflite model by itself here it is (tested in colab):
> 
> ```
> !wget https://github.com/google/mediapipe/raw/master/mediapipe/models/iris_landmark.tflite
> import tensorflow as tf
> import numpy as np
> 
> model_file = 'iris_landmark.tflite'
> interpreter = tf.lite.Interpreter(
>     model_path=model_file)
> interpreter.allocate_tensors()
> 
> input_details = interpreter.get_input_details()
> output_details = interpreter.get_output_details()
> 
> print(f'input_details {input_details}')
> print(f'output_details {output_details}')
> 
> # check the type of the input tensor
> floating_model = input_details[0]['dtype'] == np.float32
> print(f'is floating model: {floating_model}')
> 
> # NxHxWxC, H:1, W:2
> height = input_details[0]['shape'][1]
> width = input_details[0]['shape'][2]
> img = np.random.randint(0, 256, (1,64,64,3), np.uint8)
> 
> if floating_model:
>   img = (np.float32(img) - 0.0) / 255.0 # normalization from tflite_converter_calculator
> 
> interpreter.set_tensor(input_details[0]['index'], img)
> 
> interpreter.invoke()
> 
> output_eyes = interpreter.get_tensor(output_details[0]['index'])[0]
> output_iris = interpreter.get_tensor(output_details[1]['index'])[0]
> 
> print(f'output iris landmarks x {output_iris[::3]/64.0}') # tflite_tensors_to_landmarks_calculator
> print(f'output iris landmarks y {output_iris[1::3]/64.0}')
> print(f'output iris landmarks z {output_iris[2::3]}')
> 
> print(f'output eye landmarks x {output_eyes[::3]/64.0}')
> print(f'output eye landmarks y {output_eyes[1::3]/64.0}')
> print(f'output eye landmarks z {output_eyes[2::3]}')
> ```

Sir I have a couple of doubts in this.
1.  Why are the x and y coordinates of  output iris landmarks and output eye landmarks normalised using 64 and why the same is not done to z axis?
2.  Also in the model card it is mentioned that for output iris landmarks ([1,15] ) the output is  5 2D landmarks , 1 for pupil center and 4 for iris contour ,i.e, 1*(x,y,z) pupil center + 4 * (x,y,z) iris contours
Which rank in the output represents the pupil center?

**Comment by Monibsediqi** [comment]
@sandeep22-v A quick answer to your second question. The first rank represents the pupil center. You can also visualize each iris point individually to figure out the iris center point.

**Comment by sgowroji** [comment]
Hi @shortcipher3, Could you please respond if you still looking answers for the above queries. Thanks!
